File: java/leetcode_grind/Day618.java
package leetcode_grind;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;

public class Day618 {
    static class Pair<F, S> {
        F key;
        S value;

        Pair(F f, S s) {
            key = f;
            value = s;
        }

        F getKey() {
            return key;
        }

        S getValue() {
            return value;
        }
    }

    // https://leetcode.com/problems/minimum-cost-to-convert-string-i/description/?envType=daily-question&envId=2024-07-27
    static class Solution1 {
        public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
            List<int[]>[] adjacencyList = new List[26];
            for (int i = 0; i < 26; i++) {
                adjacencyList[i] = new ArrayList<>();
            }
            int conversionCount = original.length;
            for (int i = 0; i < conversionCount; i++) {
                adjacencyList[original[i] - 'a'].add(new int[] { changed[i] - 'a', cost[i] });
            }

            long[][] minConversionCosts = new long[26][26];
            for (int i = 0; i < 26; i++) {
                minConversionCosts[i] = dijkstra(i, adjacencyList);
            }

            long totalCost = 0;
            int stringLength = source.length();
            for (int i = 0; i < stringLength; i++) {
                if (source.charAt(i) != target.charAt(i)) {
                    long charConversionCost = minConversionCosts[source.charAt(i) - 'a'][target.charAt(i) - 'a'];
                    if (charConversionCost == -1)
                        return -1;
                    totalCost += charConversionCost;
                }
            }
            return totalCost;
        }

        long[] dijkstra(int startChar, List<int[]>[] adjacencyList) {
            PriorityQueue<Pair<Long, Integer>> priorityQueue = new PriorityQueue<>(
                    (e1, e2) -> Long.compare(e1.getKey(), e2.getKey()));
            priorityQueue.add(new Pair<>(0L, startChar));
            long[] minCosts = new long[26];
            Arrays.fill(minCosts, -1L);

            while (!priorityQueue.isEmpty()) {
                Pair<Long, Integer> currentPair = priorityQueue.poll();
                long currentCost = currentPair.getKey();
                int currentChar = currentPair.getValue();
                if (minCosts[currentChar] != -1L && minCosts[currentChar] < currentCost)
                    continue;

                for (int[] conversion : adjacencyList[currentChar]) {
                    int targetChar = conversion[0];
                    long conversionCost = conversion[1];
                    long newTotalCost = currentCost + conversionCost;
                    if (minCosts[targetChar] == -1L || newTotalCost < minCosts[targetChar]) {
                        minCosts[targetChar] = newTotalCost;
                        priorityQueue.add(new Pair<>(newTotalCost, targetChar));
                    }
                }
            }
            return minCosts;
        }
    }

}

File: src/leetcode_grind/day_618.rs
pub fn minimum_cost(
    source: String,
    target: String,
    original: Vec<char>,
    changed: Vec<char>,
    cost: Vec<i32>,
) -> i64 {
    let source = source.chars().collect::<Vec<_>>();
    let target = target.chars().collect::<Vec<_>>();

    let mut total_cost = 0;
    let mut min_cost = vec![vec![i32::MAX as i64; 26]; 26];

    for i in 0..original.len() {
        let start_char = original[i] as usize - 'a' as usize;
        let end_char = changed[i] as usize - 'a' as usize;

        min_cost[start_char][end_char] = min_cost[start_char][end_char].min(cost[i] as i64);
    }

    for k in 0..26 {
        for i in 0..26 {
            for j in 0..26 {
                min_cost[i][j] = min_cost[i][j].min(min_cost[i][k] + min_cost[k][j]);
            }
        }
    }

    for i in 0..source.len() {
        if source[i] == target[i] {
            continue;
        }
        let source_char = source[i] as usize - 'a' as usize;
        let target_char = target[i] as usize - 'a' as usize;

        if min_cost[source_char][target_char] >= i32::MAX as i64 {
            return -1;
        }
        total_cost += min_cost[source_char][target_char];
    }

    total_cost
}
